### PR TITLE
🏗️ Nextauth JWT & DB user role protected routes

### PR DESCRIPTION
**Pull request by shon-button**
Addresses [Ticket 265](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/265)

## 🚧 Changes
    - Modify client/app/api/auth/[...nextauth]/route.ts
    - Modify client/middlewares/withAuthorization.tsx
    - Modify client/app/middleware.ts
    - Modify client/app/(authenticated)/dashboard
    - Modify client/app/data/dashboard
    - Add protected routes client/app/(authenticated)/:
        - bceidbusiness/industry_user
        - bceidbusiness/industry_user_admin
        - idir/cas_admin
        - idir/cas_analyst
    - Add AppRole to bc_obps/registration/schema.py
    - Add GET endpoint for user role:  bc_obps/registration/api/user.py
    - Add my IDIR guid user to fixture: bc_obps/registration/fixtures/user.json

  

## 🚀 Impact:
   - App route protection implementation uses app directory to structure protected page routes based on NextAuth identity provider and user role as folders.
   - Each protected page has its own folder structure based on NextAuth JWT identity provider, NextAuth JWT user role, and a dashboard folder (used for dynamic breadcrumbs);   e.g. `app/(authenticated)/idir/cas_ admin/dashboard/operations/Page.tsx`
   - Access control logic is managed using Next.js middleware and NextAuth.js authentication JWT token.  The middleware intercepts requests, checks for a valid user session, and extracts user information from the JWT token. Based on JWT properties of identity_provider and role, the middleware  dynamically rewrites the request URL base request to the appropriate folder structure e.g. `http://localhost:3000/dashboard/operations` request maps to `http://localhost:3000/idir/cas_admin/dashboard/operations` for user `idir\cas_admin`
   - Using NextAuth JWT properties based static app directory to structure protected page routes and middleware logic to build dynamic responses allows for separation of concerns, maintainability and flexibility to delete or add roles without modifying existing access control logic.
 

## 📝 Notes:
- Using dynamic route creation based on JWT token properties attempts to align access control with the user's identity and role attributes which can then accommodate changes to user roles without requiring modifying existing access control logic. Static app directory protected routes can be added or removed without making any other changes. 
- One caveat is that page route components may be similar for multiple roles e.g. `app/(authenticated)/idir/cas_admin/dashboard/operations` and  `app/(authenticated)/idir/cas_analyst/dashboard/operations`; so, common components are used to share between protected page routes to reduce repetition.

## 🛠️ To Do:
- The BCEID tole is not finalized [321](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/321)
 - For outstanding FE PRs the legacy file structure remain but, new features will need to be integrated into protected routes as required.   [322](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/322)

##  🔬 Testing:

1.  From terminal command, start the app development server:
   ```
cd client && yarn dev
   ```
2. In your web browser, navigate to `http://localhost:3000`
3. You should be directed to the anonymous landing page
image
4. Click "Program Administrator Log In"
5. You should be directed to IDIR login screen
image
6. Login using your IDIR account
7. You should be redirected to the idir/cas_pending dashboard
image
8. Click your IDIR login name in the right hand header
9. Your user session information should display your GUID information
image
10. Add a user to `bc_obps/registration/fixtures/user.json` reflecting your IDIR GUID having `user_guid=your GUID information` and `app_role=cas_admin`
11.  Follow the [developer environment setup](https://github.com/bcgov/cas-registration/blob/develop/docs/developer-environment-setup.md#backend-environment-setup) to load fixtures  with your IDIR user data and start the API server
```
cd bc_obps
poetry shell
make loadfixtures
make run
   ```
12.  In your web browser, navigate to `http://localhost:3000`
13. Click "Log Out"
14. You should be directed to the anonymous landing page
15. Click "Program Administrator Log In"
16.  You should be directed to IDIR login screen
17. Login using your IDIR account
18. You should be redirected to the idir/cas_admin dashboard
image
19. Click your IDIR login name in the right hand header
20. Your user session information should display
image
21. Your Role should reflect  `cas_admin` and your IP should be `idir`
22. Click the breadcrumb "Dashboard"
23. You should be redirected to the `idir/cas_admin` dashboard
24. Click all dashboard navigation tiles
25. You should be directed to the route page
26. Modify the browser URL to an app path accessible to `idir\cas_admin` e.g. `http://localhost:3000/idir/cas_admin/dashboard/operations`
27. You should be redirected to URL:  `http://localhost:3000/dashboard/operations` retaining the expected breadcrumb structure
28. Modify the browser URL to an app path **NOT** accessible to `idir\cas_admin` e.g.  `http://localhost:3000/idir/cas_analyst/dashboard/operations`
29. You should be redirected to "Page not found"

### CODE REVIEW

**Comment by andrea-williams** [comment]
I haven't looked at the actual code yet, but just reviewed the existing routes-user roles mapping we discussed this morning. Here's the changes I'm seeing:

app/(authenticated)/idir/cas_admin: 
          - delete the create route within /operations (internal staff won't need to create operations themselves)
          - delete /select-operator/request-access/confirm and /received routes (these only apply to external users). We will eventually be adding a route for cas_admin to review requests for access, but that user story hasn't been built out yet

app/(authenticated)/idir/cas_analyst:
          - delete the create route within /operations (internal staff won't need to create operations themselves)
          - delete /select-operator/request-access/confirm and /received routes. I'm not sure yet whether cas_analysts will be reviewing access requests from external users - will confirm with PO

app/(authenticated)/bceid/industry_user:
          - delete the create route within /operations (only industry_admin will be allowed to create operations for their operator)

File: client/app/data/dashboard/industry_user.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[
+  {
+    "title": "My Operator",
+    "content": "View or update your operator information.",
```
Comment (dleard) [inline client/app/data/dashboard/industry_user.json:4]: Non-admin industry users will only have view privileges for the My Operator page

**Review by dleard — CHANGES_REQUESTED** [review]
This is awesome work! I followed the instructions you laid out and everything worked as expected.

I found a spelling issue that got propagated a few times, and there is still a bit of a question mark around how best to apply and enforce admin vs non admin permissions for the industry_user which I'm happy to chat about tomorrow.

I agree with Andrea's comment above that some of the pages defined here won't be necessary based on differing workflows for the idir/bceid users.

**Comment by shon-button** [comment]
Re: cleanup role routes and spelling

**AW:**
app/(authenticated)/idir/cas_admin:
- delete /operations/create route
- delete /select-operator/request-access/confirm 
- delete /select-operator/request-access/received 

app/(authenticated)/idir/cas_analyst:
- delete /operations/create route
- delete /select-operator/request-access/confirm 
- delete /select-operator/request-access/received 

app/(authenticated)/bceid/industry_user:
- delete /operations/create route

**DL**
- Operater=Operator
app/(authenticated)/bceid/industry_user:
- delete /operations route (Non-admin industry users will only have view privileges for the My Operator page)

**Pull request closed without merging**